{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Documenting how to compute specificity for RNAseq datasets\n",
    "- In this document I want to explain step by step how specificity is computed. \n",
    "- Because these codes are not yet generalizable, meaning that there are a lot of hard-coded path. Therefore, I would suggest:\n",
    "    - first, go over this notebook to understand how these statistics are\n",
    "    - second, copy over relevant codes to a fresh python script so that you can scale it up easily\n",
    "\n",
    "## Specificity\n",
    "- There are 3 ways (that I am aware of) to compute specificity\n",
    "    - Script from Rachel (original script is `single_cell_MAGMA_files_v2.py` but I have modified this to `compute_sumstat_magma.py`). For convenience, I will call this method `spec_vanilla`\n",
    "    - EWCE: method used in Skene papers. I will call this `spec_ewce`\n",
    "    - CELLEX: method from Timshel et al. I will call this `spec_cellex`\n",
    "- Below I will attempt to demonstrate how to calculate these"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### spec_vanilla\n",
    "- TL;DR: \n",
    "    - In this method, for each gene, the specifity value is computed for each cell type. Thus, for each gene, the sum of the specificity values should sum up to 1\n",
    "    - First, for each gene, mean expression is computed for each cell type. Here, gene expression value has already been converted to per million. Then, the specificity is calculated by dividing the mean for each cell type by the sum of mean expression across all cell types\n",
    "    - In this method, for each gene, the mean expression (per million) for each cell type is 0 if it's less than 1. The rationale is that if the mean expression for that cell type is less than 1 (per million), then it is lowly expressed and should be removed. \n",
    "        - However, this is somehow assumed that the gene where the mean expression for a specific cell type is low, it means that it is \"bad\" and should be removed. But what if it is actually lowly expressed? \n",
    "    - Then, after the specificity for each gene for each cell type is done, to get the top X% specificity\n",
    "        - For each cell type, remove genes where specificity is 0, then get the top X% (this is done in Rachel's script)\n",
    "            - Consequence: the number of top X% expressed genes per cell type is going to differ between cell types. This is because the number of genes that are removed due to specificity being 0 is different per cell type."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- In the following code I will walk through an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import anndata\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- First, read in and clean the h5ad file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\tph205\\AppData\\Local\\anaconda3\\envs\\jupyter_env\\lib\\site-packages\\anndata\\_core\\anndata.py:1832: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Originally, adata has 71852 rows (cells) and 36515 columns (genes)\n",
      "After cleaning, adata has 71852 rows (cells) and 24617 columns (genes)\n"
     ]
    }
   ],
   "source": [
    "adata_path = \"scrna/41_Siletti_Cerebellum.CBV_Human_2022/41_Siletti_Cerebellum.CBV_Human_2022.h5ad\" #this is the h5ad file after preprocessing\n",
    "adata_original = anndata.read(adata_path)\n",
    "print(f\"Originally, adata has {adata_original.shape[0]} rows (cells) and {adata_original.shape[1]} columns (genes)\")\n",
    "\n",
    "# because the scRNAseq data contains genes that don't have ensemble, it is cleaner to create a filtered adata where:\n",
    "# genes that were not convertible to ensembl\n",
    "adata_original.var_names_make_unique()\n",
    "gene_list = adata_original.var[adata_original.var[\"ensembl\"].notnull()].index.values.tolist() #this obtains a list of gene in symbol that has an ensemble conversion\n",
    "adata = adata_original[:, gene_list].copy() #subset based on the genes (keep the genes if it was converted to ensemble successfully)\n",
    "\n",
    "# now, adata.X is the matrix in numpy format where row is cell ID and column is the gene\\\n",
    "print(f\"After cleaning, adata has {adata.shape[0]} rows (cells) and {adata.shape[1]} columns (genes)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Committed oligodendrocyte precursor', 'Oligodendrocyte', 'Oligodendrocyte precursor', 'Splatter', 'Upper rhombic lip', ..., 'Bergmann glia', 'Ependymal', 'Vascular', 'Fibroblast', 'Microglia']\n",
       "Length: 13\n",
       "Categories (13, object): ['Astrocyte', 'Bergmann glia', 'Cerebellar inhibitory', 'Committed oligodendrocyte precursor', ..., 'Oligodendrocyte precursor', 'Splatter', 'Upper rhombic lip', 'Vascular']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# adata.obs.columns\n",
    "adata.obs[\"supercluster_term\"].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Second, compute specificity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The mean dataframe has 13 rows (cell types) and 24617 columns (genes)\n",
      "View the dataframe means_cell_counts_pM\n",
      "                                     ENSG00000284678  ENSG00000167995   \n",
      "Committed oligodendrocyte precursor         5.413892        88.919067  \\\n",
      "Oligodendrocyte                             0.900931        52.096020   \n",
      "Oligodendrocyte precursor                   0.417204        19.098175   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper rhombic lip                           0.198289         1.025848   \n",
      "Cerebellar inhibitory                       0.312472         0.523617   \n",
      "Miscellaneous                               0.000000         0.460915   \n",
      "Astrocyte                                   0.000000        26.481335   \n",
      "Bergmann glia                               0.000000         2.707252   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Vascular                                    0.000000         3.799043   \n",
      "Fibroblast                                  0.000000         3.451866   \n",
      "Microglia                                   2.586608        12.717940   \n",
      "\n",
      "                                     ENSG00000204655  ENSG00000253807   \n",
      "Committed oligodendrocyte precursor       207.840424      1616.560669  \\\n",
      "Oligodendrocyte                           711.774658       456.298828   \n",
      "Oligodendrocyte precursor                   7.468224       125.204872   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper rhombic lip                           1.897831         3.852722   \n",
      "Cerebellar inhibitory                       1.162573         4.760777   \n",
      "Miscellaneous                               7.858354         1.151087   \n",
      "Astrocyte                                   8.159776        55.050003   \n",
      "Bergmann glia                               1.646641        14.669123   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Vascular                                    4.781935         2.953710   \n",
      "Fibroblast                                  6.997016         2.810086   \n",
      "Microglia                                   3.740892         2.735217   \n",
      "\n",
      "                                     ENSG00000169247  \n",
      "Committed oligodendrocyte precursor        22.990683  \n",
      "Oligodendrocyte                           313.670349  \n",
      "Oligodendrocyte precursor                   1.159492  \n",
      "Splatter                                    0.000000  \n",
      "Upper rhombic lip                           0.598138  \n",
      "Cerebellar inhibitory                       0.619504  \n",
      "Miscellaneous                               0.185358  \n",
      "Astrocyte                                   4.407663  \n",
      "Bergmann glia                               0.615803  \n",
      "Ependymal                                   0.000000  \n",
      "Vascular                                    8.499635  \n",
      "Fibroblast                                  0.000000  \n",
      "Microglia                                   2.211884  \n",
      "The low filter dataframe has 13 rows (cell types) and 24617 columns (genes)\n",
      "View the dataframe low filter\n",
      "                                     ENSG00000284678  ENSG00000167995   \n",
      "Committed oligodendrocyte precursor                1                1  \\\n",
      "Oligodendrocyte                                    0                1   \n",
      "Oligodendrocyte precursor                          0                1   \n",
      "Splatter                                           0                0   \n",
      "Upper rhombic lip                                  0                1   \n",
      "Cerebellar inhibitory                              0                0   \n",
      "Miscellaneous                                      0                0   \n",
      "Astrocyte                                          0                1   \n",
      "Bergmann glia                                      0                1   \n",
      "Ependymal                                          0                0   \n",
      "Vascular                                           0                1   \n",
      "Fibroblast                                         0                1   \n",
      "Microglia                                          1                1   \n",
      "\n",
      "                                     ENSG00000204655  ENSG00000253807   \n",
      "Committed oligodendrocyte precursor                1                1  \\\n",
      "Oligodendrocyte                                    1                1   \n",
      "Oligodendrocyte precursor                          1                1   \n",
      "Splatter                                           0                0   \n",
      "Upper rhombic lip                                  1                1   \n",
      "Cerebellar inhibitory                              1                1   \n",
      "Miscellaneous                                      1                1   \n",
      "Astrocyte                                          1                1   \n",
      "Bergmann glia                                      1                1   \n",
      "Ependymal                                          0                0   \n",
      "Vascular                                           1                1   \n",
      "Fibroblast                                         1                1   \n",
      "Microglia                                          1                1   \n",
      "\n",
      "                                     ENSG00000169247  \n",
      "Committed oligodendrocyte precursor                1  \n",
      "Oligodendrocyte                                    1  \n",
      "Oligodendrocyte precursor                          1  \n",
      "Splatter                                           0  \n",
      "Upper rhombic lip                                  0  \n",
      "Cerebellar inhibitory                              0  \n",
      "Miscellaneous                                      0  \n",
      "Astrocyte                                          1  \n",
      "Bergmann glia                                      0  \n",
      "Ependymal                                          0  \n",
      "Vascular                                           1  \n",
      "Fibroblast                                         0  \n",
      "Microglia                                          1  \n",
      "The specificity dataframe has 13 rows (cell types) and 24617 columns (genes)\n",
      "View the dataframe specificity\n",
      "                                     ENSG00000284678  ENSG00000167995   \n",
      "Committed oligodendrocyte precursor         0.676694         0.422827  \\\n",
      "Oligodendrocyte                             0.000000         0.247726   \n",
      "Oligodendrocyte precursor                   0.000000         0.090815   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper rhombic lip                           0.000000         0.004878   \n",
      "Cerebellar inhibitory                       0.000000         0.000000   \n",
      "Miscellaneous                               0.000000         0.000000   \n",
      "Astrocyte                                   0.000000         0.125924   \n",
      "Bergmann glia                               0.000000         0.012873   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Vascular                                    0.000000         0.018065   \n",
      "Fibroblast                                  0.000000         0.016414   \n",
      "Microglia                                   0.323306         0.060476   \n",
      "\n",
      "                                     ENSG00000204655  ENSG00000253807   \n",
      "Committed oligodendrocyte precursor         0.215752         0.707142  \\\n",
      "Oligodendrocyte                             0.738870         0.199602   \n",
      "Oligodendrocyte precursor                   0.007753         0.054769   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper rhombic lip                           0.001970         0.001685   \n",
      "Cerebellar inhibitory                       0.001207         0.002083   \n",
      "Miscellaneous                               0.008158         0.000504   \n",
      "Astrocyte                                   0.008470         0.024081   \n",
      "Bergmann glia                               0.001709         0.006417   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Vascular                                    0.004964         0.001292   \n",
      "Fibroblast                                  0.007263         0.001229   \n",
      "Microglia                                   0.003883         0.001196   \n",
      "\n",
      "                                     ENSG00000169247  \n",
      "Committed oligodendrocyte precursor         0.065141  \n",
      "Oligodendrocyte                             0.888736  \n",
      "Oligodendrocyte precursor                   0.003285  \n",
      "Splatter                                    0.000000  \n",
      "Upper rhombic lip                           0.000000  \n",
      "Cerebellar inhibitory                       0.000000  \n",
      "Miscellaneous                               0.000000  \n",
      "Astrocyte                                   0.012488  \n",
      "Bergmann glia                               0.000000  \n",
      "Ependymal                                   0.000000  \n",
      "Vascular                                    0.024082  \n",
      "Fibroblast                                  0.000000  \n",
      "Microglia                                   0.006267  \n"
     ]
    }
   ],
   "source": [
    "ct_colname = \"supercluster_term\"\n",
    "# first normalize\n",
    "# normalize counts per cell (in place)\n",
    "sc.pp.normalize_total(adata, target_sum=1e6)\n",
    "\n",
    "# define the genes in the dataset\n",
    "genes = adata.var[\"ensembl\"].to_list()\n",
    "\n",
    "# define the available cell types in the dataset\n",
    "cts = adata.obs[ct_colname].dropna().unique()\n",
    "\n",
    "# before log-transforming, we determine a filtering on a minimum of 1 transcript per million per celltype\n",
    "# start with computing the mean cpM per cell type\n",
    "means_cell_counts_pM = pd.DataFrame(data=None, index=cts, columns=genes, dtype=float)  # rows is the cell type; columns is the gene\n",
    "for ct in cts:\n",
    "    means_cell_counts_pM.loc[ct, :] = adata[adata.obs[ct_colname] == ct, :].X.mean(0)\n",
    "\n",
    "print(f\"The mean dataframe has {means_cell_counts_pM.shape[0]} rows (cell types) and {means_cell_counts_pM.shape[1]} columns (genes)\")\n",
    "print(\"View the dataframe means_cell_counts_pM\")\n",
    "print(means_cell_counts_pM.iloc[:, : 5])\n",
    "\n",
    "\n",
    "# the low_filter matrix contains zeros for genes that have < 1 count pM in a certain cell type\n",
    "# this matrix will be used to multiply the same data when it is log-transformed\n",
    "low_filter = (1 - (means_cell_counts_pM < 1))  # 0 is <1, 1 if >1\n",
    "print(f\"The low filter dataframe has {low_filter.shape[0]} rows (cell types) and {low_filter.shape[1]} columns (genes)\")\n",
    "print(\"View the dataframe low filter\")\n",
    "print(low_filter.iloc[:, : 5])\n",
    "\n",
    "# specificity values\n",
    "# set NaNs (i.e. no expression in any of the cells/celltypes) in the mean matrix per cell type to zero\n",
    "spec_cell_counts_pM = ((means_cell_counts_pM.mul(low_filter)) / (means_cell_counts_pM.mul(low_filter)).sum(axis=0)).fillna(0)\n",
    "# more explanation: first, for each gene, if the mean expression for a cell type is less than 1, it will be first converted to 0. Then, specificity is computed by dividing mean expression for a cell type by the sum mean expression across all cell types. In the case of NA due to dividing by 0, convert to 0 afterwards. \n",
    "print(f\"The specificity dataframe has {spec_cell_counts_pM.shape[0]} rows (cell types) and {spec_cell_counts_pM.shape[1]} columns (genes)\")\n",
    "print(\"View the dataframe specificity\")\n",
    "print(spec_cell_counts_pM.iloc[:, : 5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Sanity check: examining the dataframe for specifity, we can confirm that for each gene, the sum of specificity across all cell types is summed up to 1. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Now onto how to obtain: for each gene, the top 10 percent genes that are specific for this cell type? And how to define the background genes to be used for LDSC-SEG? \n",
    "- What Rachel implemented originally in the script `single_cell_MAGMA_files_v2.py` is: \n",
    "    - for each cell type, remove the genes where specificity is 0.\n",
    "    - for each cell type, obtain the top 10% of genes that are specifiic for that cell type using a list of genes where specificity value of 0 have been removed\n",
    "    - What this means is that the number of top 10% specific genes for each cell type is going to differ (this is because the number of genes where specificity value of 0 is going to differ between cell types)\n",
    "    - Then, **how to obtain a list of background genes?** The background genes are defined as the genes where the sum of specificity across all cell type is greater than 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Committed oligodendrocyte precursor\n",
      "Index(['ENSG00000284678', 'ENSG00000167995', 'ENSG00000204655',\n",
      "       'ENSG00000253807', 'ENSG00000071991', 'ENSG00000160951',\n",
      "       'ENSG00000158859', 'ENSG00000167123', 'ENSG00000130158',\n",
      "       'ENSG00000167565',\n",
      "       ...\n",
      "       'ENSG00000181031', 'ENSG00000141582', 'ENSG00000117009',\n",
      "       'ENSG00000093072', 'ENSG00000261485', 'ENSG00000156697',\n",
      "       'ENSG00000259642', 'ENSG00000122025', 'ENSG00000160570',\n",
      "       'ENSG00000177548'],\n",
      "      dtype='object', length=1261)\n",
      "Oligodendrocyte\n",
      "Index(['ENSG00000167995', 'ENSG00000204655', 'ENSG00000253807',\n",
      "       'ENSG00000169247', 'ENSG00000071991', 'ENSG00000167641',\n",
      "       'ENSG00000158859', 'ENSG00000167123', 'ENSG00000168314',\n",
      "       'ENSG00000105695',\n",
      "       ...\n",
      "       'ENSG00000086288', 'ENSG00000232684', 'ENSG00000105643',\n",
      "       'ENSG00000180539', 'ENSG00000076258', 'ENSG00000162729',\n",
      "       'ENSG00000236463', 'ENSG00000158406', 'ENSG00000213440',\n",
      "       'ENSG00000180739'],\n",
      "      dtype='object', length=1413)\n",
      "Oligodendrocyte precursor\n",
      "Index(['ENSG00000283518', 'ENSG00000234781', 'ENSG00000236714',\n",
      "       'ENSG00000228639', 'ENSG00000256637', 'ENSG00000117501',\n",
      "       'ENSG00000152953', 'ENSG00000151892', 'ENSG00000198077',\n",
      "       'ENSG00000230102',\n",
      "       ...\n",
      "       'ENSG00000196456', 'ENSG00000105694', 'ENSG00000204979',\n",
      "       'ENSG00000172927', 'ENSG00000231966', 'ENSG00000183092',\n",
      "       'ENSG00000160293', 'ENSG00000162843', 'ENSG00000246084',\n",
      "       'ENSG00000235532'],\n",
      "      dtype='object', length=1503)\n",
      "Splatter\n",
      "Index(['ENSG00000255794', 'ENSG00000054803', 'ENSG00000185847',\n",
      "       'ENSG00000106565', 'ENSG00000261514', 'ENSG00000139865',\n",
      "       'ENSG00000122585', 'ENSG00000205442', 'ENSG00000182674',\n",
      "       'ENSG00000172497',\n",
      "       ...\n",
      "       'ENSG00000143369', 'ENSG00000141052', 'ENSG00000141753',\n",
      "       'ENSG00000172725', 'ENSG00000111481', 'ENSG00000184261',\n",
      "       'ENSG00000163382', 'ENSG00000162551', 'ENSG00000125740',\n",
      "       'ENSG00000106266'],\n",
      "      dtype='object', length=456)\n",
      "Upper rhombic lip\n",
      "Index(['ENSG00000168070', 'ENSG00000038945', 'ENSG00000255364',\n",
      "       'ENSG00000135253', 'ENSG00000253819', 'ENSG00000230426',\n",
      "       'ENSG00000197261', 'ENSG00000254456', 'ENSG00000225539',\n",
      "       'ENSG00000162761',\n",
      "       ...\n",
      "       'ENSG00000184956', 'ENSG00000079462', 'ENSG00000183092',\n",
      "       'ENSG00000122025', 'ENSG00000184709', 'ENSG00000109684',\n",
      "       'ENSG00000204653', 'ENSG00000172575', 'ENSG00000148291',\n",
      "       'ENSG00000281406'],\n",
      "      dtype='object', length=1456)\n",
      "Cerebellar inhibitory\n",
      "Index(['ENSG00000197085', 'ENSG00000248211', 'ENSG00000176956',\n",
      "       'ENSG00000138653', 'ENSG00000166569', 'ENSG00000006116',\n",
      "       'ENSG00000164082', 'ENSG00000169213', 'ENSG00000165970',\n",
      "       'ENSG00000181195',\n",
      "       ...\n",
      "       'ENSG00000228810', 'ENSG00000129749', 'ENSG00000204979',\n",
      "       'ENSG00000248810', 'ENSG00000125510', 'ENSG00000184992',\n",
      "       'ENSG00000226822', 'ENSG00000145365', 'ENSG00000253521',\n",
      "       'ENSG00000167077'],\n",
      "      dtype='object', length=1469)\n",
      "Miscellaneous\n",
      "Index(['ENSG00000164483', 'ENSG00000165061', 'ENSG00000174482',\n",
      "       'ENSG00000006128', 'ENSG00000175779', 'ENSG00000269113',\n",
      "       'ENSG00000112218', 'ENSG00000187498', 'ENSG00000104327',\n",
      "       'ENSG00000130635',\n",
      "       ...\n",
      "       'ENSG00000142765', 'ENSG00000160211', 'ENSG00000128340',\n",
      "       'ENSG00000204525', 'ENSG00000246228', 'ENSG00000167633',\n",
      "       'ENSG00000221957', 'ENSG00000005844', 'ENSG00000122122',\n",
      "       'ENSG00000246526'],\n",
      "      dtype='object', length=1306)\n",
      "Astrocyte\n",
      "Index(['ENSG00000257585', 'ENSG00000006611', 'ENSG00000228639',\n",
      "       'ENSG00000138798', 'ENSG00000147655', 'ENSG00000283503',\n",
      "       'ENSG00000203685', 'ENSG00000102468', 'ENSG00000182463',\n",
      "       'ENSG00000197410',\n",
      "       ...\n",
      "       'ENSG00000234719', 'ENSG00000119943', 'ENSG00000233993',\n",
      "       'ENSG00000229647', 'ENSG00000111676', 'ENSG00000188404',\n",
      "       'ENSG00000163346', 'ENSG00000170915', 'ENSG00000115339',\n",
      "       'ENSG00000236188'],\n",
      "      dtype='object', length=1462)\n",
      "Bergmann glia\n",
      "Index(['ENSG00000234781', 'ENSG00000224184', 'ENSG00000257986',\n",
      "       'ENSG00000155511', 'ENSG00000151834', 'ENSG00000165092',\n",
      "       'ENSG00000178031', 'ENSG00000279082', 'ENSG00000147509',\n",
      "       'ENSG00000182256',\n",
      "       ...\n",
      "       'ENSG00000197696', 'ENSG00000160293', 'ENSG00000127666',\n",
      "       'ENSG00000182183', 'ENSG00000162843', 'ENSG00000132819',\n",
      "       'ENSG00000270629', 'ENSG00000217643', 'ENSG00000158406',\n",
      "       'ENSG00000243926'],\n",
      "      dtype='object', length=1492)\n",
      "Ependymal\n",
      "Index(['ENSG00000110975', 'ENSG00000184471', 'ENSG00000163071',\n",
      "       'ENSG00000137473', 'ENSG00000105479', 'ENSG00000160401',\n",
      "       'ENSG00000165084', 'ENSG00000165164', 'ENSG00000197057',\n",
      "       'ENSG00000169064',\n",
      "       ...\n",
      "       'ENSG00000230873', 'ENSG00000115423', 'ENSG00000226823',\n",
      "       'ENSG00000144031', 'ENSG00000156140', 'ENSG00000153347',\n",
      "       'ENSG00000166596', 'ENSG00000236532', 'ENSG00000236756',\n",
      "       'ENSG00000203780'],\n",
      "      dtype='object', length=250)\n",
      "Vascular\n",
      "Index(['ENSG00000130158', 'ENSG00000167565', 'ENSG00000172197',\n",
      "       'ENSG00000129538', 'ENSG00000197301', 'ENSG00000099715',\n",
      "       'ENSG00000115290', 'ENSG00000144366', 'ENSG00000175183',\n",
      "       'ENSG00000183853',\n",
      "       ...\n",
      "       'ENSG00000189149', 'ENSG00000197019', 'ENSG00000068831',\n",
      "       'ENSG00000104998', 'ENSG00000255987', 'ENSG00000138172',\n",
      "       'ENSG00000186827', 'ENSG00000211689', 'ENSG00000135048',\n",
      "       'ENSG00000134954'],\n",
      "      dtype='object', length=1522)\n",
      "Fibroblast\n",
      "Index(['ENSG00000172197', 'ENSG00000198121', 'ENSG00000236714',\n",
      "       'ENSG00000204291', 'ENSG00000115461', 'ENSG00000134853',\n",
      "       'ENSG00000115290', 'ENSG00000258913', 'ENSG00000248211',\n",
      "       'ENSG00000144366',\n",
      "       ...\n",
      "       'ENSG00000155962', 'ENSG00000112303', 'ENSG00000257595',\n",
      "       'ENSG00000076258', 'ENSG00000160683', 'ENSG00000222383',\n",
      "       'ENSG00000261801', 'ENSG00000238160', 'ENSG00000247596',\n",
      "       'ENSG00000181104'],\n",
      "      dtype='object', length=1511)\n",
      "Microglia\n",
      "Index(['ENSG00000284678', 'ENSG00000238057', 'ENSG00000186417',\n",
      "       'ENSG00000223749', 'ENSG00000172985', 'ENSG00000079819',\n",
      "       'ENSG00000038945', 'ENSG00000011201', 'ENSG00000148680',\n",
      "       'ENSG00000227879',\n",
      "       ...\n",
      "       'ENSG00000122877', 'ENSG00000274210', 'ENSG00000184730',\n",
      "       'ENSG00000274750', 'ENSG00000158869', 'ENSG00000130429',\n",
      "       'ENSG00000198286', 'ENSG00000247774', 'ENSG00000186517',\n",
      "       'ENSG00000101096'],\n",
      "      dtype='object', length=1495)\n"
     ]
    }
   ],
   "source": [
    "for ct in spec_cell_counts_pM.index:\n",
    "    print(ct)\n",
    "    sorted = spec_cell_counts_pM.loc[ct].sort_values(ascending=False)\n",
    "    sorted=sorted[sorted>0]\n",
    "    numexpressed = len(sorted)\n",
    "    maxincl=sorted[round(0.1*numexpressed)-1]\n",
    "    new = spec_cell_counts_pM.columns[spec_cell_counts_pM.loc[ct] >= maxincl]\n",
    "    print(new)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- As seen from above, the number of genes that are specific to each cell type is different between cell types. \n",
    "  \n",
    "**Overall thoughts**\n",
    "- This is probably ok to compute it this way for now and we will make sure to explain what we did"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Now print out the top 10% specific genes to the output file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to obtain the top X percent genes that are specific for a cell type\n",
    "import os\n",
    "def get_top_perc(ctmatrix,perc, ldsc_outdir, magma_outdir, method):\n",
    "        \"\"\" ctmatrix is a pandas dataframe containing a parameter of interest; rows (celltypes) and columns (genes)\n",
    "            the top \"perc\" percentage of genes for each celltype will be written to the outputfile \n",
    "        \"\"\" \n",
    "        # Initialize outfile for magma\n",
    "        magma_outfile = open(os.path.join(magma_outdir, \"spec_\" + str(method) + \"_top10.txt\"), \"w\")\n",
    "        for ct in ctmatrix.index:\n",
    "            sorted = ctmatrix.loc[ct].sort_values(ascending=False)\n",
    "            # remove the nonzero values\n",
    "            sorted=sorted[sorted>0]\n",
    "            numexpressed = len(sorted)\n",
    "            maxincl=sorted[round(perc*numexpressed)-1]\n",
    "            ct_fmt = ct.replace(' ',\"_\")\n",
    "            print(ct)\n",
    "            print(ct_fmt)\n",
    "            # write outfile for ldsc\n",
    "            ct_outfile = open(os.path.join(ldsc_outdir, ct_fmt + \".txt\"), \"w\")\n",
    "            ct_outfile.write('\\n'.join(ctmatrix.columns[ctmatrix.loc[ct] >= maxincl]))\n",
    "            ct_outfile.close()\n",
    "            # write outfile for magma\n",
    "            magma_out = [ct_fmt]\n",
    "            for i in ctmatrix.columns[ctmatrix.loc[ct] >= maxincl]:\n",
    "                magma_out.append(i)\n",
    "            print(\"\\t\".join(magma_out), file=magma_outfile)\n",
    "        \n",
    "        # add a line with all the genes that are expressed in this dataset:\n",
    "        all_outfile = open(os.path.join(ldsc_outdir, \"All.txt\"), \"w\")\n",
    "        all_outfile.write('\\n'.join(ctmatrix.columns[ctmatrix.sum(axis=0) > 0]))\n",
    "        all_outfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Committed oligodendrocyte precursor\n",
      "Committed_oligodendrocyte_precursor\n",
      "Oligodendrocyte\n",
      "Oligodendrocyte\n",
      "Oligodendrocyte precursor\n",
      "Oligodendrocyte_precursor\n",
      "Splatter\n",
      "Splatter\n",
      "Upper rhombic lip\n",
      "Upper_rhombic_lip\n",
      "Cerebellar inhibitory\n",
      "Cerebellar_inhibitory\n",
      "Miscellaneous\n",
      "Miscellaneous\n",
      "Astrocyte\n",
      "Astrocyte\n",
      "Bergmann glia\n",
      "Bergmann_glia\n",
      "Ependymal\n",
      "Ependymal\n",
      "Vascular\n",
      "Vascular\n",
      "Fibroblast\n",
      "Fibroblast\n",
      "Microglia\n",
      "Microglia\n"
     ]
    }
   ],
   "source": [
    "get_top_perc(spec_cell_counts_pM, 0.1, \"ldsc/spec_vanilla/41_Siletti_Cerebellum.CBV_Human_2022/genelists\", \"magma/spec_vanilla_top10/41_Siletti_Cerebellum.CBV_Human_2022/gene_covar\", \"vanilla\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# save for running magma linear\n",
    "- Because the file format for ldsc is different from magma, I need to save it differently to be compatible to use with magma"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "spec_cell_counts_pM.index = [w.replace(' ', '_') for w in spec_cell_counts_pM.index.values]\n",
    "spec_cell_counts_pM_t = spec_cell_counts_pM.T\n",
    "spec_cell_counts_pM_t.index.name = \"GENE\"\n",
    "spec_cell_counts_pM_t.reset_index(inplace=True)\n",
    "spec_cell_counts_pM_out = spec_cell_counts_pM_t.drop_duplicates(subset=[\"GENE\"])\n",
    "spec_cell_counts_pM_out.to_csv(os.path.join(\"magma/spec_vanilla_linear/41_Siletti_Cerebellum.CBV_Human_2022/gene_covar/spec_vanilla_linear.txt\"), sep=\"\\t\", index=False) # save to a file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### spec_ewce\n",
    "- Skene et al. 2018 (Reference (https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6477180/): `It was calculated using the generate.celltype.data function of the EWCE package` in the method section) computed specificity using an R package called EWCE\n",
    "- Step 1: run EWCE. \n",
    "    - Use the Rscript `process_scrna_ewce.R`. Example usage in snakemake (this is just an example and one needs to adapt accordingly):\n",
    "    \n",
    "    ```\n",
    "    rule run_ewce:\n",
    "        input:\n",
    "            h5ad = os.path.join(config[\"processed_h5ad_dir\"], \"{h5ad_id}_processed_magma_celltyping.h5ad\")\n",
    "        output:\n",
    "            os.path.join(config[\"processed_h5ad_dir\"], \"{h5ad_id}_\" + \"siletti.rda\")\n",
    "        params:\n",
    "            outdir = config[\"processed_h5ad_dir\"],\n",
    "            file_prefix = \"{h5ad_id}\",\n",
    "            groupName = \"siletti\"\n",
    "        shell:\n",
    "            \"\"\"\n",
    "            Rscript process_scrna_ewce.R --h5ad {input.h5ad} --outdir {params.outdir} --file_prefix {params.file_prefix} --groupName {params.groupName}\n",
    "            \"\"\"\n",
    "    ```\n",
    "- Step 2: Convert to text file\n",
    "    - Use the Rscript `process_ewce.R` (currently hard coding for 1 example, #TODO: generalize this script)\n",
    "    - Example output of this script for specificity\n",
    "    \n",
    "    ```\n",
    "    GENE    Astrocyte       Bergmann_glia   Cerebellar_inhibitory   Committed_oligodendrocyte_precursor     Ependymal     Fibroblast      Microglia       Miscellaneous   Oligodendrocyte Oligodendrocyte_precursor       Splatter      Upper_rhombic_lip       Vascular\n",
    "    ENSG00000121410 0.0193375341490014      0.0178245768825862      0.0261292653835096      0       0       0.0840939042241185    0.0959529061791269      0       0.395047887373253       0.0103135545556687      0       0.14467409748486      0.206626258236432\n",
    "    ```\n",
    "    - Note that when implementing ewce, it computes both the specificity and mean. In addition, in the `process_ewce.R` script, I'm also computing the average as well. Here is an example:\n",
    "    ```\n",
    "    GENE    Astrocyte       Bergmann_glia   Cerebellar_inhibitory   Committed_oligodendrocyte_precursor     Ependymal     Fibroblast      Microglia       Miscellaneous   Oligodendrocyte Oligodendrocyte_precursor       Splatter      Upper_rhombic_lip       Vascular        Average\n",
    "    ENSG00000121410 0.00492610837438424     0.00619990926962044     0.0207334861429464      0       0       0.0173410404624277    0.0153738644304682      0       0.122331879733762       0.00362506473329881     0       0.0437523585429199    0.0411255411255411      0.0211853271396437\n",
    "    ENSG00000148584 0       0.000453651897777106    0.0738695561575328      0       0       0       0       0.0857142857142857    0.00160661005278862     0.00207146556188503     0       0.000843525938422606    0       0.0126583919478994\n",
    "    ENSG00000175899 0.00985221674876847     0.0562528353243611      0.0365701187747447      0.0333333333333333\n",
    "    0     0.0346820809248555      2.14325646401118        0.242857142857143       0.0406242827633693      0.0515277058518902    1       0.0429754267575307      4.33116883116883        0.617161572193539\n",
    "    ``` \n",
    "- Step 3: Obtain the genes with top X% in specificity\n",
    "    - Note: Here I would like to use the same Python function as above (function `get_top_perc`). However, in theory, if you don't want to have to run multiple script, you can update the R script above (`process_ewce.R`) to get the top X% instead\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<bound method NDFrame.head of                  Astrocyte  Bergmann_glia  Cerebellar_inhibitory   \n",
      "GENE                                                               \n",
      "ENSG00000121410   0.019338       0.017825               0.026129  \\\n",
      "ENSG00000148584   0.000000       0.005407               0.385931   \n",
      "ENSG00000175899   0.000962       0.004023               0.001146   \n",
      "ENSG00000166535   0.480268       0.072990               0.068351   \n",
      "ENSG00000256069   0.017136       0.078206               0.025598   \n",
      "...                    ...            ...                    ...   \n",
      "ENSG00000203995   0.000000       0.056879               0.177211   \n",
      "ENSG00000162378   0.059627       0.076378               0.160735   \n",
      "ENSG00000159840   0.071161       0.033241               0.076343   \n",
      "ENSG00000074755   0.104901       0.073885               0.091331   \n",
      "ENSG00000036549   0.043547       0.061985               0.041524   \n",
      "\n",
      "                 Committed_oligodendrocyte_precursor  Ependymal  Fibroblast   \n",
      "GENE                                                                          \n",
      "ENSG00000121410                             0.000000   0.000000    0.084094  \\\n",
      "ENSG00000148584                             0.000000   0.000000    0.000000   \n",
      "ENSG00000175899                             0.001990   0.000000    0.004184   \n",
      "ENSG00000166535                             0.079479   0.000000    0.055695   \n",
      "ENSG00000256069                             0.070895   0.000000    0.000000   \n",
      "...                                              ...        ...         ...   \n",
      "ENSG00000203995                             0.000000   0.000000    0.000000   \n",
      "ENSG00000162378                             0.078219   0.000000    0.113841   \n",
      "ENSG00000159840                             0.065424   0.000000    0.137538   \n",
      "ENSG00000074755                             0.068526   0.000000    0.118048   \n",
      "ENSG00000036549                             0.104190   0.152348    0.105714   \n",
      "\n",
      "                 Microglia  Miscellaneous  Oligodendrocyte   \n",
      "GENE                                                         \n",
      "ENSG00000121410   0.095953       0.000000         0.395048  \\\n",
      "ENSG00000148584   0.000000       0.551158         0.021509   \n",
      "ENSG00000175899   0.332762       0.009371         0.003263   \n",
      "ENSG00000166535   0.004333       0.066039         0.016935   \n",
      "ENSG00000256069   0.112084       0.137449         0.141211   \n",
      "...                    ...            ...              ...   \n",
      "ENSG00000203995   0.060066       0.305159         0.066348   \n",
      "ENSG00000162378   0.086925       0.039995         0.074026   \n",
      "ENSG00000159840   0.096300       0.036240         0.036367   \n",
      "ENSG00000074755   0.117677       0.101223         0.087661   \n",
      "ENSG00000036549   0.091591       0.034268         0.143440   \n",
      "\n",
      "                 Oligodendrocyte_precursor  Splatter  Upper_rhombic_lip   \n",
      "GENE                                                                      \n",
      "ENSG00000121410                   0.010314  0.000000           0.144674  \\\n",
      "ENSG00000148584                   0.024432  0.000000           0.011563   \n",
      "ENSG00000175899                   0.003647  0.093785           0.003535   \n",
      "ENSG00000166535                   0.062207  0.000000           0.093702   \n",
      "ENSG00000256069                   0.148189  0.000000           0.076493   \n",
      "...                                    ...       ...                ...   \n",
      "ENSG00000203995                   0.045654  0.000000           0.213799   \n",
      "ENSG00000162378                   0.078120  0.000000           0.140531   \n",
      "ENSG00000159840                   0.069581  0.000000           0.146581   \n",
      "ENSG00000074755                   0.064204  0.000000           0.087935   \n",
      "ENSG00000036549                   0.094021  0.000000           0.042692   \n",
      "\n",
      "                 Vascular  \n",
      "GENE                       \n",
      "ENSG00000121410  0.206626  \n",
      "ENSG00000148584  0.000000  \n",
      "ENSG00000175899  0.541331  \n",
      "ENSG00000166535  0.000000  \n",
      "ENSG00000256069  0.192739  \n",
      "...                   ...  \n",
      "ENSG00000203995  0.074885  \n",
      "ENSG00000162378  0.091603  \n",
      "ENSG00000159840  0.231224  \n",
      "ENSG00000074755  0.084610  \n",
      "ENSG00000036549  0.084682  \n",
      "\n",
      "[22564 rows x 13 columns]>\n",
      "<bound method NDFrame.head of GENE                                 ENSG00000121410  ENSG00000148584   \n",
      "Astrocyte                                   0.019338         0.000000  \\\n",
      "Bergmann_glia                               0.017825         0.005407   \n",
      "Cerebellar_inhibitory                       0.026129         0.385931   \n",
      "Committed_oligodendrocyte_precursor         0.000000         0.000000   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.084094         0.000000   \n",
      "Microglia                                   0.095953         0.000000   \n",
      "Miscellaneous                               0.000000         0.551158   \n",
      "Oligodendrocyte                             0.395048         0.021509   \n",
      "Oligodendrocyte_precursor                   0.010314         0.024432   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper_rhombic_lip                           0.144674         0.011563   \n",
      "Vascular                                    0.206626         0.000000   \n",
      "\n",
      "GENE                                 ENSG00000175899  ENSG00000166535   \n",
      "Astrocyte                                   0.000962         0.480268  \\\n",
      "Bergmann_glia                               0.004023         0.072990   \n",
      "Cerebellar_inhibitory                       0.001146         0.068351   \n",
      "Committed_oligodendrocyte_precursor         0.001990         0.079479   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.004184         0.055695   \n",
      "Microglia                                   0.332762         0.004333   \n",
      "Miscellaneous                               0.009371         0.066039   \n",
      "Oligodendrocyte                             0.003263         0.016935   \n",
      "Oligodendrocyte_precursor                   0.003647         0.062207   \n",
      "Splatter                                    0.093785         0.000000   \n",
      "Upper_rhombic_lip                           0.003535         0.093702   \n",
      "Vascular                                    0.541331         0.000000   \n",
      "\n",
      "GENE                                 ENSG00000256069  ENSG00000184389   \n",
      "Astrocyte                                   0.017136         0.156787  \\\n",
      "Bergmann_glia                               0.078206         0.070498   \n",
      "Cerebellar_inhibitory                       0.025598         0.217176   \n",
      "Committed_oligodendrocyte_precursor         0.070895         0.000000   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.000000         0.000000   \n",
      "Microglia                                   0.112084         0.247538   \n",
      "Miscellaneous                               0.137449         0.000000   \n",
      "Oligodendrocyte                             0.141211         0.072113   \n",
      "Oligodendrocyte_precursor                   0.148189         0.059729   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper_rhombic_lip                           0.076493         0.176159   \n",
      "Vascular                                    0.192739         0.000000   \n",
      "\n",
      "GENE                                 ENSG00000128274  ENSG00000118017   \n",
      "Astrocyte                                   0.000000         0.000000  \\\n",
      "Bergmann_glia                               0.002597         0.045399   \n",
      "Cerebellar_inhibitory                       0.001830         0.023995   \n",
      "Committed_oligodendrocyte_precursor         0.000000         0.000000   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.260447         0.731797   \n",
      "Microglia                                   0.002895         0.113863   \n",
      "Miscellaneous                               0.000000         0.000000   \n",
      "Oligodendrocyte                             0.000984         0.058049   \n",
      "Oligodendrocyte_precursor                   0.000978         0.019232   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper_rhombic_lip                           0.001315         0.007665   \n",
      "Vascular                                    0.728955         0.000000   \n",
      "\n",
      "GENE                                 ENSG00000094914  ENSG00000081760  ...   \n",
      "Astrocyte                                   0.160293         0.083818  ...  \\\n",
      "Bergmann_glia                               0.097673         0.049308  ...   \n",
      "Cerebellar_inhibitory                       0.064816         0.124923  ...   \n",
      "Committed_oligodendrocyte_precursor         0.091471         0.115590  ...   \n",
      "Ependymal                                   0.000000         0.000000  ...   \n",
      "Fibroblast                                  0.080123         0.131626  ...   \n",
      "Microglia                                   0.061087         0.070893  ...   \n",
      "Miscellaneous                               0.133006         0.032015  ...   \n",
      "Oligodendrocyte                             0.098301         0.122882  ...   \n",
      "Oligodendrocyte_precursor                   0.068435         0.098987  ...   \n",
      "Splatter                                    0.000000         0.000000  ...   \n",
      "Upper_rhombic_lip                           0.060865         0.111036  ...   \n",
      "Vascular                                    0.083929         0.058922  ...   \n",
      "\n",
      "GENE                                 ENSG00000174442  ENSG00000122952   \n",
      "Astrocyte                                   0.000000         0.000000  \\\n",
      "Bergmann_glia                               0.059297         0.001939   \n",
      "Cerebellar_inhibitory                       0.044573         0.016396   \n",
      "Committed_oligodendrocyte_precursor         0.000000         0.000000   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.084962         0.000000   \n",
      "Microglia                                   0.198294         0.058351   \n",
      "Miscellaneous                               0.313419         0.691717   \n",
      "Oligodendrocyte                             0.035944         0.059496   \n",
      "Oligodendrocyte_precursor                   0.049123         0.039424   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper_rhombic_lip                           0.082539         0.035680   \n",
      "Vascular                                    0.131848         0.096997   \n",
      "\n",
      "GENE                                 ENSG00000198205  ENSG00000198455   \n",
      "Astrocyte                                   0.027986         0.067219  \\\n",
      "Bergmann_glia                               0.102556         0.065486   \n",
      "Cerebellar_inhibitory                       0.069550         0.063746   \n",
      "Committed_oligodendrocyte_precursor         0.231567         0.185398   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.040568         0.129919   \n",
      "Microglia                                   0.075745         0.045483   \n",
      "Miscellaneous                               0.032068         0.154048   \n",
      "Oligodendrocyte                             0.105121         0.059257   \n",
      "Oligodendrocyte_precursor                   0.077829         0.069994   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper_rhombic_lip                           0.126838         0.096446   \n",
      "Vascular                                    0.110172         0.063004   \n",
      "\n",
      "GENE                                 ENSG00000070476  ENSG00000203995   \n",
      "Astrocyte                                   0.125953         0.000000  \\\n",
      "Bergmann_glia                               0.100604         0.056879   \n",
      "Cerebellar_inhibitory                       0.050185         0.177211   \n",
      "Committed_oligodendrocyte_precursor         0.054194         0.000000   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.090559         0.000000   \n",
      "Microglia                                   0.099315         0.060066   \n",
      "Miscellaneous                               0.053112         0.305159   \n",
      "Oligodendrocyte                             0.095161         0.066348   \n",
      "Oligodendrocyte_precursor                   0.068405         0.045654   \n",
      "Splatter                                    0.130965         0.000000   \n",
      "Upper_rhombic_lip                           0.056748         0.213799   \n",
      "Vascular                                    0.074800         0.074885   \n",
      "\n",
      "GENE                                 ENSG00000162378  ENSG00000159840   \n",
      "Astrocyte                                   0.059627         0.071161  \\\n",
      "Bergmann_glia                               0.076378         0.033241   \n",
      "Cerebellar_inhibitory                       0.160735         0.076343   \n",
      "Committed_oligodendrocyte_precursor         0.078219         0.065424   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.113841         0.137538   \n",
      "Microglia                                   0.086925         0.096300   \n",
      "Miscellaneous                               0.039995         0.036240   \n",
      "Oligodendrocyte                             0.074026         0.036367   \n",
      "Oligodendrocyte_precursor                   0.078120         0.069581   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper_rhombic_lip                           0.140531         0.146581   \n",
      "Vascular                                    0.091603         0.231224   \n",
      "\n",
      "GENE                                 ENSG00000074755  ENSG00000036549  \n",
      "Astrocyte                                   0.104901         0.043547  \n",
      "Bergmann_glia                               0.073885         0.061985  \n",
      "Cerebellar_inhibitory                       0.091331         0.041524  \n",
      "Committed_oligodendrocyte_precursor         0.068526         0.104190  \n",
      "Ependymal                                   0.000000         0.152348  \n",
      "Fibroblast                                  0.118048         0.105714  \n",
      "Microglia                                   0.117677         0.091591  \n",
      "Miscellaneous                               0.101223         0.034268  \n",
      "Oligodendrocyte                             0.087661         0.143440  \n",
      "Oligodendrocyte_precursor                   0.064204         0.094021  \n",
      "Splatter                                    0.000000         0.000000  \n",
      "Upper_rhombic_lip                           0.087935         0.042692  \n",
      "Vascular                                    0.084610         0.084682  \n",
      "\n",
      "[13 rows x 22564 columns]>\n"
     ]
    }
   ],
   "source": [
    "# Example python code to obtain the genes with top X% in specificity\n",
    "# load in the file\n",
    "spec_ewce_orig = pd.read_csv(\"scrna/41_Siletti_Cerebellum.CBV_Human_2022/spec_ewce_linear.tsv\", sep=\"\\t\")\n",
    "spec_ewce_orig.set_index('GENE', inplace=True)\n",
    "print(spec_ewce_orig.head)\n",
    "spec_ewce = spec_ewce_orig.transpose()\n",
    "print(spec_ewce.head)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Sanity check to make sure that the sum of the columns is 1\n",
    "spec_ewce['ENSG00000175899'].sum()\n",
    "spec_ewce['ENSG00000198455'].sum()\n",
    "\n",
    "get_top_perc(spec_ewce, 0.1, \"ldsc/spec_ewce/41_Siletti_Cerebellum.CBV_Human_2022/genelists\", \"magma/spec_ewce_top10/41_Siletti_Cerebellum.CBV_Human_2022/gene_covar\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- For magma linear, copy over the file: \n",
    "```\n",
    "cp ../scrna/41_Siletti_Cerebellum.CBV_Human_2022/spec_ewce_linear.tsv spec_ewce_linear/41_Siletti_Cerebellum.CBV_Human_2022/gene_covar/\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### spec_cellex\n",
    "- cellex is a method developed and used in Timshel et al. 2020 (https://github.com/perslab/CELLEX)\n",
    "- Step 1: run cellex\n",
    "    - Note #1: because each scRNASeq data needs to be tailored, I suggest that the script used to compute specificity using cellex should also needs to be tailored and adapted\n",
    "    - Example: `process_scrna_cellex_siletti.py` (location on snellius home directory: `/gpfs/home6/tphung/projects/alc_jeanne/code/ct_enrichment/cellect/cellex/process_scrna_cellex_siletti.py`)\n",
    "    - Note #2: I have computed specificity using cellex for a few scRNAseq data already. They are stored in the shared directory on snellius: `/gpfs/work5/0/vusr0480/Processed_scRNA/data/cellex/`\n",
    "    - Example output: \n",
    "    ```\n",
    "    head 41_Siletti_Cerebellum.CBV_Human_2022.esmu_fmt.csv\n",
    "    gene,Astrocyte,Bergmann_glia,Cerebellar_inhibitory,Committed_oligodendrocyte_precursor,Ependymal,Fibroblast,Microglia,Miscellaneous,Oligodendrocyte,Oligodendrocyte_precursor,Splatter,Upper_rhombic_lip,Vascular\n",
    "    ENSG00000284678,0.0,0.0,0.0455216731859184,0.6023177951786117,0.0,0.0,0.3862554246926514,0.0,0.1257935750541277,0.0405809483126868,0.0,0.0,0.0\n",
    "    ENSG00000167995,0.2786127578454396,0.0,0.0,0.8486931953358344,0.0,0.0,0.1147825655528267,0.0,0.7639259838721524,0.3911041191936009,0.0,0.0,0.0\n",
    "    ENSG00000204655,0.0,0.0,0.0,0.7199639397931027,0.0,0.0,0.0,0.0,0.9601996237244824,0.0,0.0,0.0,0.0\n",
    "    ENSG00000253807,0.0579020801623541,0.0,0.0,0.9452305902508408,0.0,0.0,0.0,0.0,0.8377288286595166,0.4817267013403095,0.0,0.0,0.0\n",
    "    ENSG00000169247,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.9772780005252236,0.0,0.0,0.0,0.0\n",
    "    ENSG00000071991,0.0,0.0,0.0,0.6643209030420585,0.0,0.0,0.0,0.0,0.9677135629703932,0.0,0.0,0.0,0.0\n",
    "    ENSG00000238057,0.0,0.2839402620494266,0.0,0.2051677573286847,0.0,0.0,0.8302622848540593,0.0,0.3562490855241139,0.0,0.0,0.0,0.0\n",
    "    ENSG00000160951,0.0,0.0,0.0062664644368164,0.8592925359251826,0.0,0.0,0.1099966825726586,0.0,0.1692315950828765,0.0,0.0,0.0,0.2189957185631276\n",
    "    ENSG00000167641,0.0,0.0,0.0,0.2637550975168584,0.0,0.0,0.0,0.0,0.8828416228884106,0.0,0.1872113356181184,0.0,0.2979612427739169\n",
    "    ``` \n",
    "- Step 2: Obtain the genes with top X% in specificity\n",
    "    - Note: Similar to the case of `cellex_ewce` above, Here I would like to use the same Python function as above (function `get_top_perc`).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<bound method NDFrame.head of                  Astrocyte  Bergmann_glia  Cerebellar_inhibitory   \n",
      "gene                                                               \n",
      "ENSG00000284678   0.000000       0.000000               0.045522  \\\n",
      "ENSG00000167995   0.278613       0.000000               0.000000   \n",
      "ENSG00000204655   0.000000       0.000000               0.000000   \n",
      "ENSG00000253807   0.057902       0.000000               0.000000   \n",
      "ENSG00000169247   0.000000       0.000000               0.000000   \n",
      "...                    ...            ...                    ...   \n",
      "ENSG00000167633   0.000000       0.000000               0.000000   \n",
      "ENSG00000221957   0.000000       0.000000               0.000000   \n",
      "ENSG00000005844   0.000000       0.000000               0.000000   \n",
      "ENSG00000122122   0.000000       0.000000               0.000000   \n",
      "ENSG00000246526   0.029521       0.060136               0.000000   \n",
      "\n",
      "                 Committed_oligodendrocyte_precursor  Ependymal  Fibroblast   \n",
      "gene                                                                          \n",
      "ENSG00000284678                             0.602318        0.0         0.0  \\\n",
      "ENSG00000167995                             0.848693        0.0         0.0   \n",
      "ENSG00000204655                             0.719964        0.0         0.0   \n",
      "ENSG00000253807                             0.945231        0.0         0.0   \n",
      "ENSG00000169247                             0.000000        0.0         0.0   \n",
      "...                                              ...        ...         ...   \n",
      "ENSG00000167633                             0.000000        0.0         0.0   \n",
      "ENSG00000221957                             0.000000        0.0         0.0   \n",
      "ENSG00000005844                             0.000000        0.0         0.0   \n",
      "ENSG00000122122                             0.000000        0.0         0.0   \n",
      "ENSG00000246526                             0.000000        0.0         0.0   \n",
      "\n",
      "                 Microglia  Miscellaneous  Oligodendrocyte   \n",
      "gene                                                         \n",
      "ENSG00000284678   0.386255       0.000000         0.125794  \\\n",
      "ENSG00000167995   0.114783       0.000000         0.763926   \n",
      "ENSG00000204655   0.000000       0.000000         0.960200   \n",
      "ENSG00000253807   0.000000       0.000000         0.837729   \n",
      "ENSG00000169247   0.000000       0.000000         0.977278   \n",
      "...                    ...            ...              ...   \n",
      "ENSG00000167633   0.000000       0.887767         0.000000   \n",
      "ENSG00000221957   0.000000       0.937798         0.000000   \n",
      "ENSG00000005844   0.484924       0.876532         0.000000   \n",
      "ENSG00000122122   0.776257       0.907322         0.000000   \n",
      "ENSG00000246526   0.000000       0.813294         0.000000   \n",
      "\n",
      "                 Oligodendrocyte_precursor  Splatter  Upper_rhombic_lip   \n",
      "gene                                                                      \n",
      "ENSG00000284678                   0.040581       0.0           0.000000  \\\n",
      "ENSG00000167995                   0.391104       0.0           0.000000   \n",
      "ENSG00000204655                   0.000000       0.0           0.000000   \n",
      "ENSG00000253807                   0.481727       0.0           0.000000   \n",
      "ENSG00000169247                   0.000000       0.0           0.000000   \n",
      "...                                    ...       ...                ...   \n",
      "ENSG00000167633                   0.000000       0.0           0.045179   \n",
      "ENSG00000221957                   0.000000       0.0           0.062036   \n",
      "ENSG00000005844                   0.000000       0.0           0.246587   \n",
      "ENSG00000122122                   0.000000       0.0           0.000000   \n",
      "ENSG00000246526                   0.000000       0.0           0.000000   \n",
      "\n",
      "                 Vascular  \n",
      "gene                       \n",
      "ENSG00000284678       0.0  \n",
      "ENSG00000167995       0.0  \n",
      "ENSG00000204655       0.0  \n",
      "ENSG00000253807       0.0  \n",
      "ENSG00000169247       0.0  \n",
      "...                   ...  \n",
      "ENSG00000167633       0.0  \n",
      "ENSG00000221957       0.0  \n",
      "ENSG00000005844       0.0  \n",
      "ENSG00000122122       0.0  \n",
      "ENSG00000246526       0.0  \n",
      "\n",
      "[18913 rows x 13 columns]>\n",
      "<bound method NDFrame.head of gene                                 ENSG00000284678  ENSG00000167995   \n",
      "Astrocyte                                   0.000000         0.278613  \\\n",
      "Bergmann_glia                               0.000000         0.000000   \n",
      "Cerebellar_inhibitory                       0.045522         0.000000   \n",
      "Committed_oligodendrocyte_precursor         0.602318         0.848693   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.000000         0.000000   \n",
      "Microglia                                   0.386255         0.114783   \n",
      "Miscellaneous                               0.000000         0.000000   \n",
      "Oligodendrocyte                             0.125794         0.763926   \n",
      "Oligodendrocyte_precursor                   0.040581         0.391104   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper_rhombic_lip                           0.000000         0.000000   \n",
      "Vascular                                    0.000000         0.000000   \n",
      "\n",
      "gene                                 ENSG00000204655  ENSG00000253807   \n",
      "Astrocyte                                   0.000000         0.057902  \\\n",
      "Bergmann_glia                               0.000000         0.000000   \n",
      "Cerebellar_inhibitory                       0.000000         0.000000   \n",
      "Committed_oligodendrocyte_precursor         0.719964         0.945231   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.000000         0.000000   \n",
      "Microglia                                   0.000000         0.000000   \n",
      "Miscellaneous                               0.000000         0.000000   \n",
      "Oligodendrocyte                             0.960200         0.837729   \n",
      "Oligodendrocyte_precursor                   0.000000         0.481727   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper_rhombic_lip                           0.000000         0.000000   \n",
      "Vascular                                    0.000000         0.000000   \n",
      "\n",
      "gene                                 ENSG00000169247  ENSG00000071991   \n",
      "Astrocyte                                   0.000000         0.000000  \\\n",
      "Bergmann_glia                               0.000000         0.000000   \n",
      "Cerebellar_inhibitory                       0.000000         0.000000   \n",
      "Committed_oligodendrocyte_precursor         0.000000         0.664321   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.000000         0.000000   \n",
      "Microglia                                   0.000000         0.000000   \n",
      "Miscellaneous                               0.000000         0.000000   \n",
      "Oligodendrocyte                             0.977278         0.967714   \n",
      "Oligodendrocyte_precursor                   0.000000         0.000000   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper_rhombic_lip                           0.000000         0.000000   \n",
      "Vascular                                    0.000000         0.000000   \n",
      "\n",
      "gene                                 ENSG00000238057  ENSG00000160951   \n",
      "Astrocyte                                   0.000000         0.000000  \\\n",
      "Bergmann_glia                               0.283940         0.000000   \n",
      "Cerebellar_inhibitory                       0.000000         0.006266   \n",
      "Committed_oligodendrocyte_precursor         0.205168         0.859293   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.000000         0.000000   \n",
      "Microglia                                   0.830262         0.109997   \n",
      "Miscellaneous                               0.000000         0.000000   \n",
      "Oligodendrocyte                             0.356249         0.169232   \n",
      "Oligodendrocyte_precursor                   0.000000         0.000000   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper_rhombic_lip                           0.000000         0.000000   \n",
      "Vascular                                    0.000000         0.218996   \n",
      "\n",
      "gene                                 ENSG00000167641  ENSG00000158859  ...   \n",
      "Astrocyte                                   0.000000         0.000000  ...  \\\n",
      "Bergmann_glia                               0.000000         0.000000  ...   \n",
      "Cerebellar_inhibitory                       0.000000         0.000000  ...   \n",
      "Committed_oligodendrocyte_precursor         0.263755         0.236593  ...   \n",
      "Ependymal                                   0.000000         0.000000  ...   \n",
      "Fibroblast                                  0.000000         0.000000  ...   \n",
      "Microglia                                   0.000000         0.000000  ...   \n",
      "Miscellaneous                               0.000000         0.000000  ...   \n",
      "Oligodendrocyte                             0.882842         0.959198  ...   \n",
      "Oligodendrocyte_precursor                   0.000000         0.000000  ...   \n",
      "Splatter                                    0.187211         0.000000  ...   \n",
      "Upper_rhombic_lip                           0.000000         0.000000  ...   \n",
      "Vascular                                    0.297961         0.000000  ...   \n",
      "\n",
      "gene                                 ENSG00000142765  ENSG00000160211   \n",
      "Astrocyte                                   0.000000         0.000000  \\\n",
      "Bergmann_glia                               0.000000         0.000000   \n",
      "Cerebellar_inhibitory                       0.000000         0.204825   \n",
      "Committed_oligodendrocyte_precursor         0.000000         0.000000   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.000000         0.000000   \n",
      "Microglia                                   0.000000         0.000000   \n",
      "Miscellaneous                               0.642934         0.591267   \n",
      "Oligodendrocyte                             0.000000         0.000000   \n",
      "Oligodendrocyte_precursor                   0.000000         0.000000   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper_rhombic_lip                           0.556757         0.000000   \n",
      "Vascular                                    0.000000         0.000000   \n",
      "\n",
      "gene                                 ENSG00000128340  ENSG00000204525   \n",
      "Astrocyte                                   0.000000         0.237442  \\\n",
      "Bergmann_glia                               0.000000         0.304195   \n",
      "Cerebellar_inhibitory                       0.000000         0.000000   \n",
      "Committed_oligodendrocyte_precursor         0.000000         0.000000   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.000000         0.112854   \n",
      "Microglia                                   0.452409         0.304868   \n",
      "Miscellaneous                               0.961561         0.672934   \n",
      "Oligodendrocyte                             0.000000         0.000000   \n",
      "Oligodendrocyte_precursor                   0.000000         0.371603   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper_rhombic_lip                           0.000000         0.000000   \n",
      "Vascular                                    0.068303         0.688688   \n",
      "\n",
      "gene                                 ENSG00000246228  ENSG00000167633   \n",
      "Astrocyte                                   0.000000         0.000000  \\\n",
      "Bergmann_glia                               0.000000         0.000000   \n",
      "Cerebellar_inhibitory                       0.193562         0.000000   \n",
      "Committed_oligodendrocyte_precursor         0.193931         0.000000   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.000000         0.000000   \n",
      "Microglia                                   0.000000         0.000000   \n",
      "Miscellaneous                               0.671955         0.887767   \n",
      "Oligodendrocyte                             0.000000         0.000000   \n",
      "Oligodendrocyte_precursor                   0.000000         0.000000   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper_rhombic_lip                           0.000000         0.045179   \n",
      "Vascular                                    0.000000         0.000000   \n",
      "\n",
      "gene                                 ENSG00000221957  ENSG00000005844   \n",
      "Astrocyte                                   0.000000         0.000000  \\\n",
      "Bergmann_glia                               0.000000         0.000000   \n",
      "Cerebellar_inhibitory                       0.000000         0.000000   \n",
      "Committed_oligodendrocyte_precursor         0.000000         0.000000   \n",
      "Ependymal                                   0.000000         0.000000   \n",
      "Fibroblast                                  0.000000         0.000000   \n",
      "Microglia                                   0.000000         0.484924   \n",
      "Miscellaneous                               0.937798         0.876532   \n",
      "Oligodendrocyte                             0.000000         0.000000   \n",
      "Oligodendrocyte_precursor                   0.000000         0.000000   \n",
      "Splatter                                    0.000000         0.000000   \n",
      "Upper_rhombic_lip                           0.062036         0.246587   \n",
      "Vascular                                    0.000000         0.000000   \n",
      "\n",
      "gene                                 ENSG00000122122  ENSG00000246526  \n",
      "Astrocyte                                   0.000000         0.029521  \n",
      "Bergmann_glia                               0.000000         0.060136  \n",
      "Cerebellar_inhibitory                       0.000000         0.000000  \n",
      "Committed_oligodendrocyte_precursor         0.000000         0.000000  \n",
      "Ependymal                                   0.000000         0.000000  \n",
      "Fibroblast                                  0.000000         0.000000  \n",
      "Microglia                                   0.776257         0.000000  \n",
      "Miscellaneous                               0.907322         0.813294  \n",
      "Oligodendrocyte                             0.000000         0.000000  \n",
      "Oligodendrocyte_precursor                   0.000000         0.000000  \n",
      "Splatter                                    0.000000         0.000000  \n",
      "Upper_rhombic_lip                           0.000000         0.000000  \n",
      "Vascular                                    0.000000         0.000000  \n",
      "\n",
      "[13 rows x 18913 columns]>\n"
     ]
    }
   ],
   "source": [
    "# Example python code to obtain the genes with top X% in specificity\n",
    "# load in the file\n",
    "spec_cellex_orig = pd.read_csv(\"scrna/41_Siletti_Cerebellum.CBV_Human_2022/41_Siletti_Cerebellum.CBV_Human_2022.esmu_fmt.csv\")\n",
    "spec_cellex_orig.set_index('gene', inplace=True)\n",
    "print(spec_cellex_orig.head)\n",
    "spec_cellex = spec_cellex_orig.transpose()\n",
    "print(spec_cellex.head)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.9998340082613578\n",
      "1.6080424367931252\n",
      "0.9329457677807221\n",
      "2.6925855842211823\n"
     ]
    }
   ],
   "source": [
    "#Sanity check to make sure that the sum of the columns is 1\n",
    "print(spec_cellex['ENSG00000221957'].sum())\n",
    "print(spec_cellex['ENSG00000005844'].sum())\n",
    "print(spec_cellex['ENSG00000167633'].sum())\n",
    "print(spec_cellex['ENSG00000204525'].sum())\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- This is very strange. I'm not sure why the specificity per gene does not add up to 1. #TODO: INVESTIGATE CELLEX"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "get_top_perc(spec_cellex, 0.1, \"ldsc/spec_cellex/41_Siletti_Cerebellum.CBV_Human_2022/genelists\", \"magma/spec_cellex_top10/41_Siletti_Cerebellum.CBV_Human_2022/gene_covar\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- In addition, there are around half of the genes that are specific when using `cellex` as compared to when using `vanilla` or `ewce`. #TODO: INVESTIGATE THIS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Save for running magma linear"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<bound method NDFrame.head of                   gene  Astrocyte  Bergmann_glia  Cerebellar_inhibitory   \n",
      "0      ENSG00000284678   0.000000       0.000000               0.045522  \\\n",
      "1      ENSG00000167995   0.278613       0.000000               0.000000   \n",
      "2      ENSG00000204655   0.000000       0.000000               0.000000   \n",
      "3      ENSG00000253807   0.057902       0.000000               0.000000   \n",
      "4      ENSG00000169247   0.000000       0.000000               0.000000   \n",
      "...                ...        ...            ...                    ...   \n",
      "18908  ENSG00000167633   0.000000       0.000000               0.000000   \n",
      "18909  ENSG00000221957   0.000000       0.000000               0.000000   \n",
      "18910  ENSG00000005844   0.000000       0.000000               0.000000   \n",
      "18911  ENSG00000122122   0.000000       0.000000               0.000000   \n",
      "18912  ENSG00000246526   0.029521       0.060136               0.000000   \n",
      "\n",
      "       Committed_oligodendrocyte_precursor  Ependymal  Fibroblast  Microglia   \n",
      "0                                 0.602318        0.0         0.0   0.386255  \\\n",
      "1                                 0.848693        0.0         0.0   0.114783   \n",
      "2                                 0.719964        0.0         0.0   0.000000   \n",
      "3                                 0.945231        0.0         0.0   0.000000   \n",
      "4                                 0.000000        0.0         0.0   0.000000   \n",
      "...                                    ...        ...         ...        ...   \n",
      "18908                             0.000000        0.0         0.0   0.000000   \n",
      "18909                             0.000000        0.0         0.0   0.000000   \n",
      "18910                             0.000000        0.0         0.0   0.484924   \n",
      "18911                             0.000000        0.0         0.0   0.776257   \n",
      "18912                             0.000000        0.0         0.0   0.000000   \n",
      "\n",
      "       Miscellaneous  Oligodendrocyte  Oligodendrocyte_precursor  Splatter   \n",
      "0           0.000000         0.125794                   0.040581       0.0  \\\n",
      "1           0.000000         0.763926                   0.391104       0.0   \n",
      "2           0.000000         0.960200                   0.000000       0.0   \n",
      "3           0.000000         0.837729                   0.481727       0.0   \n",
      "4           0.000000         0.977278                   0.000000       0.0   \n",
      "...              ...              ...                        ...       ...   \n",
      "18908       0.887767         0.000000                   0.000000       0.0   \n",
      "18909       0.937798         0.000000                   0.000000       0.0   \n",
      "18910       0.876532         0.000000                   0.000000       0.0   \n",
      "18911       0.907322         0.000000                   0.000000       0.0   \n",
      "18912       0.813294         0.000000                   0.000000       0.0   \n",
      "\n",
      "       Upper_rhombic_lip  Vascular  \n",
      "0               0.000000       0.0  \n",
      "1               0.000000       0.0  \n",
      "2               0.000000       0.0  \n",
      "3               0.000000       0.0  \n",
      "4               0.000000       0.0  \n",
      "...                  ...       ...  \n",
      "18908           0.045179       0.0  \n",
      "18909           0.062036       0.0  \n",
      "18910           0.246587       0.0  \n",
      "18911           0.000000       0.0  \n",
      "18912           0.000000       0.0  \n",
      "\n",
      "[18913 rows x 14 columns]>\n"
     ]
    }
   ],
   "source": [
    "spec_cellex_orig = pd.read_csv(\"scrna/41_Siletti_Cerebellum.CBV_Human_2022/41_Siletti_Cerebellum.CBV_Human_2022.esmu_fmt.csv\")\n",
    "# print(spec_cellex_orig.head)\n",
    "spec_cellex_orig.rename(columns = {'gene':'GENE'}, inplace=True)\n",
    "spec_cellex_orig = spec_cellex_orig.drop_duplicates(subset=[\"GENE\"])\n",
    "spec_cellex_orig.to_csv(os.path.join(\"magma/spec_cellex_linear/41_Siletti_Cerebellum.CBV_Human_2022/gene_covar/spec_cellex_linear.txt\"), sep=\"\\t\", index=False) # save to a file)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
