# FUMA_Celltype_cmd
This repo hosts codes for running FUMA Cell Type on the command line
- FUMA Cell Type (originally written by Kyoko) is implemented on FUMA for general use 
- The problem is that it is difficult to scale it to run for more traits and with new scRNAseq datasets
- Here, I am adapting Kyoko's script to run FUMA Cell Type on the command line with the following updates:
  - Currently only run step 1 (magma gene property). Step 2 (determining whether cell types are dependent or independent within the same dataset) might be implemented later. Step 3 is probably not necessary.
  - a snakemake workflow
  - run magma gene property for each GWAS and for each scRNAseq
  - calculate adjusted pvalue for within dataset and across datasets (across all the datasets selected)

## Initial documentation
- Documenting how to use the scripts to run FUMA Cell Type on the command line
- Demo on lisa for 1 trait and 1 scRNAseq dataset
- #TO-ADD: how to scale this up to submit as an sbatch job on snellius for efficiency
### Setup
- Demo folder on lisa: `/project/prjstphung/FUMA_CellType_Step1/`
  - There are 3 subfolders: `data`, `code`, and `output`
  - `data`:
    - create 2 subfolders: 
      - `gene_results`: 
        - subfolder: `protein_coding_ensemble` (format: {gene_region}_{gene_convention}): in this folder is the file: `AD_protein_coding_ensemble_magma.genes.raw`
          - This is generated by Rachel Preprocessing GWAS's pipeline (#TODO: link the info here)
          - Format: {trait}_{gene_region}_{gene_convention}_magma.genes.raw where {gene_region} is `protein_coding` or `all` and {gene_convention} could be `ensemble` or `entrez` (for now, this could be updated later)
      - `gene_covar`:
        - subfolder example: `id1_Allen_MCA_2019`. This is the internal id of the scRNAseq dataset
          - In this folder is: `means_cell_log_counts_pM_convert.tsv` (this is the output from script: https://github.com/ctg-single-cell/master_scripts/blob/main/Misc/compute_sumstat_magma.py)
- Summary of folder/data organization:
  - In the working directory, create 3 folders: `data`, `code`, and `output`
  - In the `data` folder, create 2 folders: 
    - `gene_results`: create subfolders following format {gene_region}_{gene_convention} such as `protein_coding_ensemble` or `all_ensemble` or `protein_entrez` 
      - then, place the file generated by Rachel such as `AD_protein_coding_ensemble_magma.genes.raw` (convention is: {trait}_{gene_region}_{gene_convention}_magma.genes.raw). Can also create a symlink instead of copy over
    - `gene_covar`: create subfolders where the name of the subfolder is the internal name of the scRNAseq data. Then, put the processed (mean) scRNAseq file here (can also create a symlink)
  - `code`: make sure that these 3 files are in this folder
    - `config.json`
    - `fuma_celltype_step1.smk`
    - `calc_adj_pval_magma_step1.R`
### Run instructions:
- Make sure to have snakemake installed
- The snakemake file `fuma_celltype_step1.smk` is written such that the file format and path have to be following the specific format (see `Setup` section above)
- Edit the file `config.json` as appropriate
- Dry run:
```
snakemake -s fuma_celltype_step1.smk -np
```
  - Examine the output to make sure that the steps are run correctly
- Actual run:
```commandline
snakemake -s fuma_celltype_step1.smk -j1
```
  - Here, this is run on the command line. This is ok for now with 1 trait and 1 scRNAseq dataset. You can also put this command in a bash script and submit to the queue. 
  - #TODO: once snellius is ready I will test how to submit this for many traits and many scRNAseq datasets. 
- Outputs:
  - The outputs are found in the `output/{trait}` folder

