---
title: "Gene locations Ensembl"
author: "Rachel Brouwer"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

Creating a list of genes and locations based on GRCh37.p13.
Genome annotations Homo_sapiens.GRCh37.87.gtf downloaded from http://ftp.ensembl.org/pub/grch37/release-104/gtf/homo_sapiens/ on 17/11/2021.



```{r}
library(stringr)
setwd("~/Documents/single_cell/gene_locations")
gene_list <- read.table("Homo_sapiens.GRCh37.87.gtf",comment.char="#",sep='\t',quote="")

# take those lines that contain the full gene
gene_list <- gene_list[gene_list[,3]=="gene",]

# get the ENSG id from column 9
gene_list$ensg <- str_split_fixed(str_split_fixed(gene_list[,9],";",2)[,1],"\"",3)[,2]

# write to file in magma gene.loc format
write.table(gene_list[,c(10,1,4,5,7)],"all_ensmbl_genes.loc", col.names=F,row.names=F,quote=F)

# select only protein coding genes by first creating a field for biotype
gene_list$biotype <- str_split_fixed(str_split_fixed(gene_list[,9],";",5)[,5],"\"",3)[,2]
# write to file in magma gene.loc format for protein coding genes only
gene_list <- gene_list[gene_list$biotype == "protein_coding",]
write.table(gene_list[,c(10,1,4,5,7)],"ensmbl_protein_coding_genes.loc", col.names=F,row.names=F,quote=F)

```